### ISSUE TITLE
How to use reward model with GPU/ 使用模型的奖励评分如何使用GPU

### ISSUE THREAD

**Issue by Guncuke** [volcengine/verl#2042]
When modifying the open-source code for training GRPO, I found that the internal naive.py file was changed. I want to introduce a model to compute the reward scores for the responses. However, I noticed that within the function in verl/workers/reward_manager/naive.py, the number of available GPUs is shown as zero. How can I modify it so that the reward function can dynamically use the GPU, just like rollout and gradient updates do?
我在修改别人论文的开源代码时，发现在训练GRPO时，修改了内部的naive.py文件，我希望可以引入一个模型，计算回答的得分，但是我发现在verl/workers/reward_manager/naive.py函数内，GPU显示数量为0，我要怎么修改，才能让这个奖励函数可以像rollout和梯度更新一样可以动态使用显卡。

我是用main_ppo.py来启动训练

```python3
from verl import DataProto
from verl.utils.reward_score import _default_compute_score
import torch
import time


class NaiveRewardManager:
    """The reward manager.
    """

    def __init__(self, tokenizer, num_examine, train_batch_size, num_generation, compute_score=None) -> None:
        self.tokenizer = tokenizer
        self.num_examine = num_examine
        self.compute_score = compute_score or _default_compute_score
        self.train_batch_size = train_batch_size
        self.num_generation = num_generation
        self.text_splitter = None
        self._text_splitter_initialized = False
    
    def _init_text_splitter(self):
        if self._text_splitter_initialized:
            return
            
        try:
            from langchain_community.embeddings import HuggingFaceBgeEmbeddings
            from langchain_experimental.text_splitter import SemanticChunker

            if torch.cuda.is_available():
                print(f"CUDA available: {torch.cuda.is_available()}")
                print(f"CUDA device count: {torch.cuda.device_count()}")
                for i in range(torch.cuda.device_count()):
                    print(f"GPU {i}: {torch.cuda.get_device_name(i)}")
                    print(f"GPU {i} memory: {torch.cuda.get_device_properties(i).total_memory / 1024**3:.2f} GB")
                print(f"Current CUDA device: {torch.cuda.current_device()}")
                device = "cuda"
            else:
                print("CUDA not available, using CPU")
                device = "cpu"

            embed_model_name = "Qwen3-Embedding-0.6B"
            model_kwargs = {'device': device}
            encode_kwargs = {'normalize_embeddings': True}
            hf = HuggingFaceBgeEmbeddings(
                model_name=embed_model_name, 
                model_kwargs=model_kwargs, 
                encode_kwargs=encode_kwargs
            )
            self.text_splitter = SemanticChunker(hf, breakpoint_threshold_amount=95)
            print(f"Text splitter initialized successfully on {device}!")
        except Exception as e:
            print(f"Warning: Failed to initialize semantic splitter: {e}")
            self.text_splitter = None
        
        self._text_splitter_initialized = True

   
    def check_correctness_length_and_segments(self, data):
        ...
```

**Comment by hyyp1** [comment]
两种做法吧：
1. 外部起一个http服务，专门算reward
2. 比较复杂，新建一个work group，实例化fsdp_worker中的ActorRolloutRefWorker类，分配GPU

**Comment by Guncuke** [comment]
![Image](https://github.com/user-attachments/assets/256076e8-bb6e-493f-96ec-27a3800eba62)
我按照文档的内容，设置了
reward_model.enable=True \
reward_model.strategy=fsdp \
但是的reward_fn内仍然没有GPU，这是为什么呢，这是main_ppo内对应的代码：
```python3
if config.reward_model.enable:
            if config.reward_model.strategy == 'fsdp':
                from verl.workers.fsdp_workers import RewardModelWorker
            elif config.reward_model.strategy == 'megatron':
                from verl.workers.megatron_workers import RewardModelWorker
            else:
                raise NotImplementedError
            role_worker_mapping[Role.RewardModel] = RewardModelWorker
            mapping[Role.RewardModel] = global_pool_id
```

**Comment by sweartoyou** [comment]
请问这个解决了吗？想请教一下您这里的解决方案

**Comment by waltonfuture** [comment]
同问

**Comment by Guncuke** [comment]
我目前是单独部署一个API，然后直接调用，用了两台机子

**Comment by Ruheng-W** [comment]
同问，难道只能另起一个机子了吗。。

**Comment by GabyUSTC** [comment]
main_ppo.py第91行，ray起服务的时候分配一个GPU资源，亲测可以解决
``` python
@ray.remote(num_cpus=1, num_gpus=1)  # please make sure main_task is not scheduled on head
class TaskRunner:
```

``` python
@register("custom")
class CustomRewardManager(AbstractRewardManager):
    """The reward manager."""

    def __init__(self, tokenizer, num_examine, compute_score=None, reward_fn_key="data_source") -> None:
        """
        Initialize the CustomRewardManager instance.

        Args:
            tokenizer: The tokenizer used to decode token IDs into text.
            num_examine: The number of batches of decoded responses to print to the console for debugging purpose.
            compute_score: A function to compute the reward score. If None, `default_compute_score` will be used.
            reward_fn_key: The key used to access the data source in the non-tensor batch data. Defaults to
                "data_source".
        """
        print("Using CustomRewardManager")
        print(f"CUDA available: {torch.cuda.is_available()}")

        self.tokenizer = tokenizer  # Store the tokenizer for decoding token IDs
        self.num_examine = num_examine  # the number of batches of decoded responses to print to the console
        self.compute_score = compute_score or default_compute_score
        self.reward_fn_key = reward_fn_key  # Store the key for accessing the data source
        self._models = {}
        self._tokenizers = {}
```

```
(TaskRunner pid=235754) Using CustomRewardManager
(TaskRunner pid=235754) CUDA available: True
```

**Comment by Guncuke** [comment]
> main_ppo.py第91行，ray起服务的时候分配一个GPU资源，亲测可以解决
> 
> @ray.remote(num_cpus=1, num_gpus=1)  # please make sure main_task is not scheduled on head
> class TaskRunner:
> @register("custom")
> class CustomRewardManager(AbstractRewardManager):
>     """The reward manager."""
> 
>     def __init__(self, tokenizer, num_examine, compute_score=None, reward_fn_key="data_source") -> None:
>         """
>         Initialize the CustomRewardManager instance.
> 
>         Args:
>             tokenizer: The tokenizer used to decode token IDs into text.
>             num_examine: The number of batches of decoded responses to print to the console for debugging purpose.
>             compute_score: A function to compute the reward score. If None, `default_compute_score` will be used.
>             reward_fn_key: The key used to access the data source in the non-tensor batch data. Defaults to
>                 "data_source".
>         """
>         print("Using CustomRewardManager")
>         print(f"CUDA available: {torch.cuda.is_available()}")
> 
>         self.tokenizer = tokenizer  # Store the tokenizer for decoding token IDs
>         self.num_examine = num_examine  # the number of batches of decoded responses to print to the console
>         self.compute_score = compute_score or default_compute_score
>         self.reward_fn_key = reward_fn_key  # Store the key for accessing the data source
>         self._models = {}
>         self._tokenizers = {}
> ```
> (TaskRunner pid=235754) Using CustomRewardManager
> (TaskRunner pid=235754) CUDA available: True
> ```

这是一个办法，但是这样子推理和训练只能使用7张卡了，我不太确定更新是否会有问题

**Comment by XuankunRong** [comment]
同问
